### PR TITLE
Implement C2C tests

### LINKED ISSUE #271: C2C feature testing and implementation
## Problem

We want to test C2C inner transaction call described in [c2c-feature-branch](https://github.com/algorand/go-algorand/tree/feature/contract-to-contract/data/transactions/logic#inner-transactions), and the discussion/design-doc is [here](https://docs.google.com/document/d/1uGM62VsN1F4aSTNg4t2q21JmuwS-OO-7bFY0VcaZcro/edit#).

We might want to have some deep-calls (a chain of inner-trans call) as testing method, which should be designed in SDK testing. The ideal result is that we can walk through inner transactions cascaded in tree-like structure.

### LINKED ISSUE #280: Remove Template Feature Related Cucumber Tests
## Problem

The integration test `template.feature` has been removed, but various steps that only belong to this test are still present.

## Solution

Remove the code associated with `template.feature` and no other running tests. In particular, here is a list of steps that can be removed:

* **Given** _a split contract with ratio <ratn> to <ratd> and minimum payment <min_pay>_
* **When** _I fund the contract account_
* **When** _I send the split transactions_
* **Given** _an HTLC contract with hash preimage "<preimage>"_
* **When** _I fund the contract account_
* **When** _I claim the algos_
* **Given** _a periodic payment contract with withdrawing window <wd_window> and period <period>_
* **When** _I claim the periodic payment_
* **Given** _a limit order contract with parameters <ratn> <ratd> <min_trade>_
* **When** _I swap assets for algos_
* **_Given_** a dynamic fee contract with amount <amt>
* **_Given_** I send the dynamic fee transactions

## Dependencies

Testing PR https://github.com/algorand/algorand-sdk-testing/pull/156 should be merged before merging the SDK PR.

## Urgency

Low

### PR DESCRIPTION
**Pull request by jasonpaulos**
- [x] Add `models.PendingTransactionInfoResponse` field to `ABIMethodResult`
- [x] Implement all `@c2c` tests
- [x] Implement new test in `@applications.verified` 
- [x] Implement new test in `@compile`
- [x] Remove templates tests -- closes #280 
- [x] Revert branch change in `run_docker.sh`

Closes #271

### CODE REVIEW

File: test/applications_integration_test.go [hunk 1]
```diff
@@ -571,6 +631,239 @@ func theAppShouldHaveReturned(commaSeparatedB64Results string) error {
...
+
+		if !bytes.Equal(reencoded, actualResult.RawReturnValue) {
+			return fmt.Errorf("The round trip result does not match the original result")
+		}
+	}
+
+	return nil
+}
+
+func checkAtomicResultAgainstValue(resultIndex int, path, expectedValue string) error {
```
Comment (jasonpaulos) [inline test/applications_integration_test.go:677]: @ahangsu this implementation might help with the Java SDK

File: test/applications_integration_test.go [hunk 2]
```diff
@@ -571,6 +636,239 @@ func theAppShouldHaveReturned(commaSeparatedB64Results string) error {
...
+
+	for i, key := range keys {
+		var value interface{}
+
+		intKey, err := strconv.Atoi(key)
+		if err == nil {
+			// key is an array index
+			genericArray, ok := genericJson.([]interface{})
+			if !ok {
+				return fmt.Errorf("Path component %d is an array index (%d), but the partent is not an array. Parent type: %s", i, intKey, reflect.TypeOf(genericJson))
```
Comment (algochoi) [inline test/applications_integration_test.go:704]:
```suggestion
				return fmt.Errorf("Path component %d is an array index (%d), but the parent is not an array. Parent type: %s", i, intKey, reflect.TypeOf(genericJson))
```

File: test/applications_integration_test.go [hunk 3]
```diff
@@ -571,6 +636,239 @@ func theAppShouldHaveReturned(commaSeparatedB64Results string) error {
...
+			}
+			if intKey < 0 || intKey >= len(genericArray) {
+				return fmt.Errorf("Path component %d is an array index (%d) outside of the parent array's range. Parent length: %d", i, intKey, len(genericArray))
+			}
+			value = genericArray[intKey]
+		} else {
+			// key is an object field
+			genericObject, ok := genericJson.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("Path component %d is an object field ('%s'), but the partent is not an object. Parent type: %s", i, key, reflect.TypeOf(genericJson))
```
Comment (algochoi) [inline test/applications_integration_test.go:714]:
```suggestion
				return fmt.Errorf("Path component %d is an object field ('%s'), but the parent is not an object. Parent type: %s", i, key, reflect.TypeOf(genericJson))
```

File: test/applications_integration_test.go [hunk 4]
```diff
@@ -571,6 +636,239 @@ func theAppShouldHaveReturned(commaSeparatedB64Results string) error {
...
+			}
+			value, ok = genericObject[key]
+			if !ok {
+				var parentFields []string
+				for field := range genericObject {
+					parentFields = append(parentFields, "'"+field+"'")
+				}
+				return fmt.Errorf("Path component %d is an object field ('%s'), but the partent does not contain the field. Partent fields are: %s", i, key, strings.Join(parentFields, ","))
```
Comment (algochoi) [inline test/applications_integration_test.go:722]:
```suggestion
				return fmt.Errorf("Path component %d is an object field ('%s'), but the parent does not contain the field. Parent fields are: %s", i, key, strings.Join(parentFields, ","))
```

**Review by algochoi — APPROVED** [review]
Nice, the implementation here definitely looks trickier than the others

File: test/applications_integration_test.go [hunk 5]
```diff
@@ -90,17 +140,11 @@ func iBuildAnApplicationTransaction(
 	}
 
 	if approvalProgram != "" {
-		approvalP, err = ioutil.ReadFile("features/resources/" + approvalProgram)
-		if err != nil {
-			return err
-		}
+		approvalP, err = readTealProgram(approvalProgram)
```
Comment (ahangsu) [inline test/applications_integration_test.go:143]: nit: do we want to abort here if this trigger an error?

**Review by ahangsu — COMMENTED** [review]
Generally LGTM, just one minor thing unclear to me

File: test/applications_integration_test.go (hunk 5 above)
Comment (jasonpaulos) [reply test/applications_integration_test.go:143]: My mistake, I did not mean to get rid of the if statement

File: test/applications_integration_test.go (hunk 2 above)
Comment (jasonpaulos) [reply test/applications_integration_test.go:704]: Thank you 😅

File: test/steps_test.go [hunk 6]
```diff
@@ -291,18 +271,6 @@ func FeatureContext(s *godog.Suite) {
...
-	s.Step(`^an HTLC contract with hash preimage "([^"]*)"$`, anHTLCContractWithHashPreimage)
-	s.Step(`^I fund the contract account$`, iFundTheContractAccount)
-	s.Step(`^I claim the algos$`, iClaimTheAlgosHTLC)
-	s.Step(`^a periodic payment contract with withdrawing window (\d+) and period (\d+)$`, aPeriodicPaymentContractWithWithdrawingWindowAndPeriod)
-	s.Step(`^I claim the periodic payment$`, iClaimThePeriodicPayment)
-	s.Step(`^a limit order contract with parameters (\d+) (\d+) (\d+)$`, aLimitOrderContractWithParameters)
-	s.Step(`^I swap assets for algos$`, iSwapAssetsForAlgos)
-	s.Step(`^a dynamic fee contract with amount (\d+)$`, aDynamicFeeContractWithAmount)
-	s.Step(`^I send the dynamic fee transactions$`, iSendTheDynamicFeeTransaction)
-	s.Step("contract test fixture", createContractTestFixture)
```
Comment (tzaffi) [inline test/steps_test.go:305]: :+1

File: test/steps_test.go [hunk 7]
```diff
@@ -1773,248 +1744,11 @@ func createRevocationTransaction(amount int) error {
...
-	secretKeyTwo := exp.PrivateKey
-	groupTxnBytes, err := templates.GetDynamicFeeTransactions(initialTxn, lsig, secretKeyTwo, uint64(params.Fee))
-	// hack to make checkTxn work
-	txn = initialTxn
-	// end hack to make checkTxn work
-	response, err := acl.SendRawTransaction(groupTxnBytes)
-	txid = response.TxID
-	return err
-}
-
```
Comment (tzaffi) [inline test/steps_test.go:2017]: 👍

File: test/steps_test.go [hunk 8]
```diff
@@ -2609,10 +2393,19 @@ func addMethodCallHelper(accountType, strOnComplete, approvalProgram, clearProgr
 		Signer:     accountTxSigner,
 	}
 
+	if useNonce {
+		methodCallParams.Note = note
```
Comment (tzaffi) [inline test/steps_test.go:2397]: 👍

**Review by tzaffi — APPROVED** [review]
LGTM

**Pull request merged**